From 7de2582f630bdd6990cd7140033c687a09aee6ec Mon Sep 17 00:00:00 2001
From: "trupti.patil" <trupti_patil@xyratex.com>
Date: Fri, 21 Sep 2012 09:58:51 +0530
Subject: [PATCH 3/3] A change so that c2_layout_update()

 - This is a change so that c2_layout_update() follows the DB5 semantics when
   a non-existing record is attempted to be updated. A new record is added and
   SUCCESS is returned in that case.
 - Inspected by Nikita against RB 1041.
---
 db/tdb/tdb.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/db/tdb/tdb.c b/db/tdb/tdb.c
index a4d50d2..d9996ac 100644
--- a/db/tdb/tdb.c
+++ b/db/tdb/tdb.c
@@ -292,11 +292,12 @@ int c2_table_update(struct c2_db_tx *tx, struct c2_db_pair *pair)
 	if (replacement != NULL) {
 		pair_lock(pair);
 		kpair = ktable_lookup(pair, &result);
-		if (result == 0) {
-			pair_tlist_add_after(kpair, replacement);
+		if (kpair != NULL) {
+			pair_tlist_add_before(kpair, replacement);
 			kpair_plank(tx, kpair);
 		} else
-			c2_free(replacement);
+			pair_tlist_add_tail(pair_list(pair), replacement);
+		result = 0;
 		pair_unlock(pair);
 	} else
 		result = -ENOMEM;
-- 
1.8.3.2

